src/hotspot/share/runtime/flags/jvmFlagLimit.hpp

```diff
@@ -0,0 +1,138 @@
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_RUNTIME_FLAGS_JVMFLAGLIMIT_HPP
+#define SHARE_RUNTIME_FLAGS_JVMFLAGLIMIT_HPP
+
+#include "runtime/flags/jvmFlag.hpp"
+
+#define ALL_LIMIT_TYPES(f)  \
+  f(int)                    \
+  f(intx)                   \
+  f(uint)                   \
+  f(uintx)                  \
+  f(uint64_t)               \
+  f(size_t)                 \
+  f(double)
+
+#define ALL_RANGE_TYPES(f) ALL_LIMIT_TYPES(f)
+#define ALL_CONSTRAINT_TYPES(f) ALL_LIMIT_TYPES(f) f(bool)
+
+template <typename T> class JVMTypedFlagLimit;
+
+class JVMFlagLimit {
+  short _constraint_func;
+  char  _phase;
+  char  _kind;
+
+ static const JVMFlagLimit* const* flagLimits;
+  static int _last_checked;
+
+protected:
+  enum {
+    HAS_RANGE = 1,
+    HAS_CONSTRAINT = 2
+  };
+
+public:
+  void* constraint_func() const;
+  char phase() const { return _phase; }
+  char kind()  const { return _kind; }
+
+  constexpr JVMFlagLimit(short func, short phase, short kind) : _constraint_func(func), _phase(phase), _kind(kind) {}
+
+  static const JVMFlagLimit* get_range(const JVMFlag* flag) {
+    return get_range_at(flag->flag_enum());
+  }
+  static const JVMFlagLimit* get_range_at(int flag_enum) {
+    const JVMFlagLimit* limit = flagLimits[flag_enum];
+    if (limit != NULL && (limit->_kind & HAS_RANGE) != 0) {
+      _last_checked = flag_enum;
+      return limit;
+    } else {
+      return NULL;
+    }
+  }
+
+  static const JVMFlagLimit* get_constraint(const JVMFlag* flag) {
+    return get_constraint_at(flag->flag_enum());
+  }
+  static const JVMFlagLimit* get_constraint_at(int flag_enum) {
+    const JVMFlagLimit* limit = flagLimits[flag_enum];
+    if (limit != NULL && (limit->_kind & HAS_CONSTRAINT) != 0) {
+      _last_checked = flag_enum;
+      return limit;
+    } else {
+      return NULL;
+    }
+  }
+  static const JVMFlag* last_checked_flag() {
+    if (_last_checked >= 0) {
+      return &JVMFlag::flags[_last_checked];
+    } else {
+      return NULL;
+    }
+  }
+
+#define AS_TYPED_LIMIT(type) inline JVMTypedFlagLimit<type>* as_ ## type() const { return (JVMTypedFlagLimit<type>*)this; }
```

Comment:
Is it possible to use static_cast here rather than C-style cast?

Reply:
I tried but got this error:

```
invalid static_cast from type 'const JVMFlagLimit*' to type 'const JVMTypedFlagLimit<unsigned int>*'
```